crates/autopilot/src/database/onchain_order_events.rs

```diff
@@ -0,0 +1,727 @@
+use super::{events::meta_to_event_index, Metrics, Postgres};
+use anyhow::{anyhow, bail, Context, Result};
+use contracts::cowswap_onchain_orders::{
+    event_data::OrderPlacement as ContractOrderPlacement, Event as ContractEvent,
+};
+use database::{
+    byte_array::ByteArray, events::EventIndex, onchain_broadcasted_orders::OnchainOrderPlacement,
+    orders::Order, PgTransaction,
+};
+use ethcontract::{Event as EthContractEvent, H160};
+use model::{
+    order::OrderUid,
+    order::{BuyTokenDestination, OrderKind, SellTokenSource},
+    signature::SigningScheme,
+    DomainSeparator,
+    {app_id::AppId, order::OrderData},
+};
+use number_conversions::u256_to_big_decimal;
+use shared::{
+    db_order_conversions::{
+        buy_token_destination_into, order_kind_into, sell_token_source_into, signing_scheme_into,
+    },
+    event_handling::EventStoring,
+    order_quoting::{OrderQuoting, QuoteSearchParameters},
+    order_validation::{
+        convert_signing_scheme_into_quote_kind, get_quote_and_check_fee,
+        is_order_outside_market_price,
+    },
+};
+use std::collections::HashMap;
+use unzip_n::unzip_n;
+
+unzip_n!(3);
+
+pub struct OnchainOrderParser<'a, T: Send + Sync, W: Send + Sync> {
+    db: Postgres,
+    quoter: Box<dyn OrderQuoting>,
+    custom_onchain_data_parser: Box<dyn CustomOnchainOrderParsing<'a, T, W>>,
+    domain_separator: DomainSeparator,
+    settlement_contract: H160,
+}
+
+impl<'a, T: Send + Sync, W: Send + Sync> OnchainOrderParser<'a, T, W> {
+    pub fn new(
+        db: Postgres,
+        quoter: Box<dyn OrderQuoting>,
+        custom_onchain_data_parser: Box<dyn CustomOnchainOrderParsing<'a, T, W>>,
+        domain_separator: DomainSeparator,
+        settlement_contract: H160,
+    ) -> Self {
+        OnchainOrderParser {
+            db,
+            quoter,
+            custom_onchain_data_parser,
+            domain_separator,
+            settlement_contract,
+        }
+    }
+}
+pub struct CustomParsedOnchaninData<T> {
+    quote_id: i64,
+    additional_data: Option<T>,
+}
+
+#[async_trait::async_trait]
+// CustomEventData generic stores the result of the custom event parsing
+// CustomEvenDataForDB generic contains the prepared data that will be appended to the database
+pub trait CustomOnchainOrderParsing<
+    'a,
+    CustomEventData: Send + Sync + Clone,
+    CustomEventDataForDB: Send + Sync,
+>: Send + Sync
+{
+    async fn append_custom_order_info_to_db(
+        &self,
+        ex: &mut PgTransaction<'a>,
+        custom_onchain_data: Vec<CustomEventDataForDB>,
+    ) -> Result<()>;
```

Comment:
```suggestion
    async fn append_custom_order_info_to_db<'a>(
        &self,
        ex: & mut PgTransaction<'a>,
        custom_onchain_data: Vec<CustomEventDataForDB>,
    ) -> Result<()>;
```
Move the lifetime into the function and remove it from the trait. This in turn allows you to remove the lifetime from several places.